### PR TITLE
Expose Array Metadata functions

### PR DESCRIPTION
**Pull request by pomadchin**
This PR is the first attempt to expose missing Array metadata functions. This piece of functionality is neccesary to allow proper reads of tiledb files [written via GDAL](https://github.com/OSGeo/gdal/blob/v3.1.0RC2/gdal/frmts/tiledb/tiledbdataset.cpp).

I hope it aligns with your development plans, and finally [this Java bindings docs page](https://docs.tiledb.com/developer/api-usage/array-metadata) would not be empty anymore.

### CODE REVIEW

**Comment by Shelnutt2** [comment]
@pomadchin thank you for the pull request. We appreciate your effort to add the metadata support. @gsvic will review the PR shortly.

A few questions/comments from me, in a more general sense. Most of the TileDB-Java API revolves around NativeArrays (as you've implemented here too). This stems from we don't often directly use the java api, but instead built ontop for spark, prestodb and others. Might I ask how you are using or planning to use the TileDB-Java API? Would it be of help to you if we introduced some functions to that accepted java objects directly instead of requiring you as the user to build NativeArrays?

For instance, In-addition/instead of returning a `NativeArray` class for the metadata fetch functions, would it be helpful to return an Object that is the native java type? There is a `NativeArray.toJavaArray()` function which will convert to a Java based array. The next thing to do would be to use the metadata length to return either a single value i.e. `Long` or if >1 value return an array `[]long`/String/appropriate type. We likely need to return a pair `<Object, class>` so you can cast the object to the appropriate type.

Similar for `putMetadata`, would you like an overloaded function which accepts a java object (with restrictions on the type) instead of a Native array?

We'd like your feedback so we can consider the improvements to make to the API. We are in the process of updating TileDB-Java for the just released TileDB 2.0, so we will be able to incorporate your comments as we make additions/changes.

**Comment by pomadchin** [comment]
Thanks for a really quick response @Shelnutt2! 

I've only started to look into the TileDB, and this is my first attemp to work with it in general. 

I'm investigating how well TileDB will work for us as a replacement for the NetCDF format. I'm pretty sure it would be more efficient (:D everything is probably more eficient than NetCDF), however I would like to get some exact relative numbers.

In general, the idea is to compare the "windowed read" performance of JVM GeoTiff / NetCDF readers vs TileDB and vs GDAL (GeoTiff / NetCDF / TileDB) interface (mb it makes not a lot of sense to use TileDB JNI directly and we can just rely on GDAL). 

Talking about the `Array Metadata API`:
I used `NativeArray` types to keep the API simple and to try to keep the current API design as it is. The `NativeArray` contains information about the underlying native data type wich is pretty convenient => it allows a bit more elegant work with types. Basing on the `NativeArray` `nativeType` it is possible to cast it into the `Java Array` of an appropriate type in a generic fashion on the user side.

Hm, mb "Java types" overloads you mentioned can accept as an input some generic type parameter to "hide" the cast from the user? In this case it could indeed simplify the life of a user. 

I use it from Scala, so sorry for my probably a bit "Scala" vision of the API and I can understand that `Objects` in the return types is a common thing in Java.

Let me show some examples of how it is possible to work with the API now:

```scala
// in Java that would be just a static cast into short[]
query.getBuffer("TDB_VALUES").asInstanceOf[Array[Short]] //> looks a bit ugly

// the cast can be hidden
query.getBuffer[Array[Short]]("TDB_VALUES")

// metadata retrieval
val md = array.getMetadata("_gdal", Datatype.TILEDB_UINT8)
val str = 
  md
    .getSecond()
    .toJavaArray()
    .asInstanceOf[Array[Short]]
    .map(_.toChar)
    .mkString

// which could probably look like 
val str = 
  array
    .getMetadata[Array[Short]]("_gdal", Datatype.TILEDB_UINT8)
    .getSecond()
    .map(_.toChar)
    .mkString

// subarray definition example
val subarray = new NativeArray(ctx, Array(zMin, zMax, rowMin, ySize - 1, colMin, xSize - 1), Datatype.TILEDB_UINT64)

// which could probably look just as 
val subarray = Array(zMin, zMax, rowMin, ySize - 1, colMin, xSize - 1)

query
  .setLayout(TILEDB_ROW_MAJOR)
  .setSubarray(subarray)
  .setBuffer("TDB_VALUES", data)
```

Probably having such kind of overloads can make user API a bit better indeed. 
It could be really convenient to have all the logic around "casting" incapuslated, so on the user side it could be possible to avoid explicit casts and just rely on type parameters. But it would still be nice to have these `NativeArray` interfaces for some not trivial user logic.

Hopefully my feedback is not ambigous

File: src/main/java/io/tiledb/java/api/Array.java [hunk 1]
```diff
@@ -615,6 +615,151 @@ public Pair getNonEmptyDomainFromName(String name) throws TileDBError {
...
+   * @return NativeArray which contains the metadata
+   * @throws TileDBError A TileDB exception
+   */
+  public NativeArray getMetadata(String key, Datatype nativeType) throws TileDBError {
+    checkIsOpen();
+
+    SWIGTYPE_p_p_void resultArrpp = tiledb.new_voidpArray(0);
+    SWIGTYPE_p_unsigned_int value_num = tiledb.new_uintp();
+    SWIGTYPE_p_tiledb_datatype_t value_type =
+        tiledb.copy_tiledb_datatype_tp(nativeType.toSwigEnum());
```
Comment (pomadchin) [inline src/main/java/io/tiledb/java/api/Array.java:633]: I thought that [nativeType is not required](https://github.com/TileDB-Inc/TileDB-Go/blob/3479a6c12c1939a44aa84b09f5140aa3cc3a4ab2/array.go#L1107-L1114) for the input.
However, `tiledb.tiledb_array_get_metadata` would throw a null pointer exception in case `value_type = tiledb.new_tiledb_datatype_tp();`

File: src/main/java/io/tiledb/java/api/Array.java [hunk 2]
```diff
@@ -615,6 +615,179 @@ public Pair getNonEmptyDomainFromName(String name) throws TileDBError {
...
+
+  /**
+   * Puts a metadata key-value item to an open array. The array must be opened in WRITE mode,
+   * otherwise the function will error out.
+   *
+   * @param key a key to assign to the input value
+   * @param value the metadata to put into the Array metadata
+   * @throws TileDBError A TileDB exception
+   */
+  public void putMetadata(String key, NativeArray value) throws TileDBError {
```
Comment (pomadchin) [inline src/main/java/io/tiledb/java/api/Array.java:778]: I'm wondering, should the following overload be a part of this PR

```java
public void putMetadata(String key, Object buffer, Class javaType) throws TileDBError {
  putMetadata(key, new NativeArray(ctx, buffer, javaType));
}
```

Or it can / should be discussed in a separate PR in case this one would pass.

**Comment by normanb** [comment]
Hi @pomadchin. Thank you for your contributions to the Java API. The `_gdal` metadata item stores the GDAL aux.xml file as you have seen. We are in the process of releasing an update to the GDAL TileDB driver that implements https://gdal.org/user/multidim_raster_data_model.html as well if that is of interest. The direct TileDB apis will be faster and easier to use though. Thank you for your inputs!

**Comment by pomadchin** [comment]
@normanb Wow, thanks for sharing that! It sounds like the GDAL 3.2 feature?

File: src/main/java/io/tiledb/java/api/Array.java (hunk 2 above)
Comment (Shelnutt2) [reply src/main/java/io/tiledb/java/api/Array.java:778]: @pomadchin let's look into this as a followup PR. We wanted to initial changes merged, as they are a good foundation to build upon.

For the overloads, is this something you are interested in contributing? If not, @gsvic can add them, we just want to coordinate.

**Comment by Shelnutt2** [comment]
@pomadchin Thanks again for the PR. If you find anything else you'd like improved or changed in the TileDB-Java API please don't hesitate to open an issue. You can also post on the [forums](https://forum.tiledb.com/) if you have more general questions or drop us an email at [hello@tiledb.com](mailto:hello@tiledb.com)

File: src/main/java/io/tiledb/java/api/Array.java (hunk 2 above)
Comment (pomadchin) [reply src/main/java/io/tiledb/java/api/Array.java:778]: Done! https://github.com/TileDB-Inc/TileDB-Java/pull/156

**Comment by pomadchin** [comment]
Thanks for merging it in! Didn't know about forums (was actually trying to find smth like that before making this PR).
